### PR TITLE
feat: typescript native bindings

### PR DESCRIPTION
**Pull request by giorgiopiatti-dfinity**
### Overview

This PR introduces new TypeScript bindings that generate more idiomatic and readable TypeScript interfaces for Candid. These bindings follow TypeScript best practices and conventions, making them more intuitive for TypeScript developers and more interpretable by AI models.

### Background

The current TypeScript bindings use JavaScript-centric patterns that, while functional, don't leverage TypeScript's type system effectively. This work builds on the foundation laid in #601 and addresses real-world usage needs from early adopters like Caffeine.

####  **More Idiomatic Type Mapping**
The new bindings provide a cleaner, more TypeScript-native representation of Candid types:

**Optionals**
<table>
<tr><th>Type</th><th>Old Binding</th><th>New Binding</th></tr>
<tr><td>Optional (bottom-most)</td><td><code>[] | [T]</code></td><td><code>T | null</code></td></tr>
<tr><td>Nested Optional</td><td><code>[] | [[] | [T]]</code></td><td><code>Optional&lt;T | null&gt;</code></td></tr>
<tr><td>Optional Record Field</td><td><code>field: [] | [string]</code></td><td><code>field?: string</code></td></tr>

</table>

_Example:_
```typescript
// Old binding
interface User {
  name: string;
  email: [] | [string];
  metadata: [] | [[] | [Record]];
}

// New binding
interface User {
  name: string;
  email?: string;
  metadata?: Option<Record>;
}
```

**Variant with no values**
old
```
type bitcoin_network = { 'mainnet' : null } | { 'testnet' : null }
```
new
```
enum bitcoin_network {
    mainnet = "mainnet",
    testnet = "testnet"
}
```

**Variant with values**
old
```
type res = {
    'Ok' : [bigint, bigint]
  } |
  {
    'Err' : {
      'error' : string,
    }
  };
  ```
  new
  ```
  type res = {
    __kind__: "Ok";
    Ok: [bigint, bigint];
} | {
    __kind__: "Err";
    Err: {
        error: string;
    };
};
```

**Service**
Service interfaces now uses method definition with arguments and typescript like interface:

_old_
```
export interface _SERVICE {
  'createTask' : ActorMethod<
    [string, [] | [Time], string, [] | [string]],
    bigint
  >,
  'getDisplayName' : ActorMethod<[], [] | [string]>,
  'getTasks' : ActorMethod<[], Array<Task>>,
  'setDisplayName' : ActorMethod<[string], undefined>,
  'updateTask' : ActorMethod<
    [bigint, string, [] | [Time], string, [] | [string], boolean],
    boolean
  >,
}

```
_new_

```
export interface backendInterface {
    createTask(title: string, dueDate: Time | null, priority: string, notes: string | null): Promise<bigint>;
    getDisplayName(): Promise<string | null>;
    getTasks(): Promise<Array<Task>>;
    setDisplayName(displayName: string): Promise<void>;
    updateTask(id: bigint, title: string, dueDate: Time | null, priority: string, notes: string | null, completed: boolean): Promise<boolean>;
}
```

### Architecture

The implementation consists of two complementary components:

1. **Interface Generator** (`ts-native-interface`): Produces clean TypeScript type definitions
2. **Wrapper Generator** (`ts-native-wrapper`): Creates a layer that handles conversion between the old typescript bindings which agent-js uses and the new type format

This dual approach ensures backward compatibility while enabling gradual migration.

#### Generate TypeScript Interface
```bash
didc bind --target ts-native-interface example.did > example.d.ts
```

#### Generate Wrapper with Conversion Logic
```bash
didc bind --target ts-native-wrapper example.did > example.ts
```

### Testing

The test suite validates type compatibility using the TypeScript compiler to ensure the wrapper correctly translates between old and new formats.

#### Run Individual Tests
```bash
cd rust/candid_parser/tests/assets/typescript_native
./test.sh ../example.did
```

#### Run All Tests
```bash
cd rust/candid_parser/tests/assets/typescript_native
./test_all.sh
```

**Test Coverage:** The suite includes comprehensive test cases for recursive types, complex nested structures, and edge cases.

### Breaking Changes

None. This PR adds new functionality without modifying existing behavior.

### Related Issues

- Continues work from #601
---

**For Reviewers:** The main implementation is in `rust/candid_parser/src/bindings/typescript_native/`. Key files to review:
- `compile.rs` - Entry point
- `new_typescript_native_types.rs` - Type conversion logic
- `compile_wrapper.rs` - Wrapper generation
- `compile_interface.rs` - Interface generation

### CODE REVIEW

File: rust/candid/Cargo.toml [hunk 1]
```diff
@@ -1,7 +1,7 @@
 [package]
 name = "candid"
 # sync with the version in `candid_derive/Cargo.toml`
-version = "0.10.16"
+version = "0.10.17"
```
Comment (giorgiopiatti-caffeinated) [inline rust/candid/Cargo.toml]:
```suggestion
version = "0.10.16"
```

File: rust/candid/Cargo.toml [hunk 2]
```diff
@@ -16,7 +16,7 @@ keywords = ["internet-computer", "idl", "candid", "dfinity"]
 include = ["src", "Cargo.toml", "LICENSE", "README.md"]
 
 [dependencies]
-candid_derive = { path = "../candid_derive", version = "=0.10.16" }
+candid_derive = { path = "../candid_derive", version = "=0.10.17" }
```
Comment (giorgiopiatti-caffeinated) [inline rust/candid/Cargo.toml]:
```suggestion
candid_derive = { path = "../candid_derive", version = "=0.10.16" }
```

File: rust/candid/src/pretty/utils.rs [hunk 3]
```diff
@@ -55,6 +55,10 @@ pub fn concat<'a, D>(docs: D, sep: &'a str) -> RcDoc<'a>
 where
     D: Iterator<Item = RcDoc<'a>>,
 {
+    let mut docs = docs.peekable();
+    if docs.peek().is_none() {
+        return RcDoc::nil();
+    }
```
Comment (giorgiopiatti-caffeinated) [inline rust/candid/src/pretty/utils.rs]:
```suggestion
```

File: rust/candid/src/pretty/utils.rs [hunk 4]
```diff
@@ -85,3 +89,16 @@ pub fn quote_ident(id: &str) -> RcDoc {
...
+    use super::*;
+
+    #[test]
+    fn concat_empty() {
+        let t = concat(vec![].into_iter(), ",")
+            .pretty(LINE_WIDTH)
+            .to_string();
+        assert_eq!(t, "");
+    }
+}
```
Comment (giorgiopiatti-caffeinated) [inline rust/candid/src/pretty/utils.rs]:
```suggestion
```

File: rust/candid_derive/Cargo.toml [hunk 5]
```diff
@@ -1,7 +1,7 @@
 [package]
 name = "candid_derive"
 # sync with the version in `candid/Cargo.toml`
-version = "0.10.16"
+version = "0.10.17"
```
Comment (giorgiopiatti-caffeinated) [inline rust/candid_derive/Cargo.toml]:
```suggestion
version = "0.10.16"
```

File: tools/didc/Cargo.toml [hunk 6]
```diff
@@ -1,6 +1,6 @@
 [package]
 name = "didc"
-version = "0.5.0"
+version = "0.5.1"
```
Comment (giorgiopiatti-caffeinated) [inline tools/didc/Cargo.toml]:
```suggestion
version = "0.5.0"
```

File: rust/bench/Cargo.lock [hunk 7]
```diff
@@ -153,7 +153,7 @@ dependencies = [
 
 [[package]]
 name = "candid"
-version = "0.10.16"
+version = "0.10.17"
```
Comment (giorgiopiatti-caffeinated) [inline rust/bench/Cargo.lock]:
```suggestion
version = "0.10.16"
```

File: rust/bench/Cargo.lock [hunk 8]
```diff
@@ -176,7 +176,7 @@ dependencies = [
 
 [[package]]
 name = "candid_derive"
-version = "0.10.16"
+version = "0.10.17"
```
Comment (giorgiopiatti-caffeinated) [inline rust/bench/Cargo.lock]:
```suggestion
version = "0.10.16"
```

File: rust/candid/Cargo.toml
Comment (ilbertt) [inline rust/candid/Cargo.toml:1]: I guess these are leftovers from the rebasing?

File: rust/candid_derive/Cargo.toml
Comment (ilbertt) [inline rust/candid_derive/Cargo.toml:1]: Same as https://github.com/dfinity/candid/pull/672#discussion_r2296692704

File: tools/didc/Cargo.toml
Comment (ilbertt) [inline tools/didc/Cargo.toml:1]: Same as https://github.com/dfinity/candid/pull/672#discussion_r2296692704

File: deny.toml
Comment (ilbertt) [inline deny.toml:1]: I'm not sure that we can allow other licenses that easily

File: rust/candid_parser/Cargo.toml [hunk 9]
```diff
@@ -41,6 +41,7 @@ num-traits = { workspace = true, optional = true }
 dialoguer = { version = "0.11", default-features = false, features = ["editor", "completion"], optional = true }
 ctrlc = { version = "3.4", optional = true }
 console = { workspace = true, optional = true }
+swc_core = { version = "18.0.0", features = ["common", "ecma_ast", "ecma_codegen"] }
```
Comment (ilbertt) [inline rust/candid_parser/Cargo.toml:44]: It looks like `swc_core` is at [v37.0.0](https://github.com/swc-project/swc/blob/940dca073dbbb7743e3dc7570f3348146861d061/crates/swc_core/Cargo.toml#L9) now. They seem to release a major version every week, so I think they don't follow semver. It may be worth it bumping it to the latest anyway

File: rust/candid_parser/src/bindings/typescript_native/comments.rs [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+        comment_text.push('\n');
+
+        Some(Comment {
+            span: DUMMY_SP,
+            kind: CommentKind::Block,
+            text: comment_text.into(),
+            // swc_core 0.80+ uses None for comments attached to no specific position
+            // If you want to attach to leading, set as Some(true)
+            // For now, we use None
+            // has_trailing_newline: false, // Only in newer swc
```
Comment (ilbertt) [inline rust/candid_parser/src/bindings/typescript_native/comments.rs:48]: Not sure if I understand this comment. It seems to point to a really old release of swc_core

File: rust/candid_parser/src/bindings/typescript_native/preamble/imports.rs [hunk 11]
```diff
@@ -0,0 +1,342 @@
...
+        type_only: false,
+        with: None,
+        phase: Default::default(),
+    };
+    module
+        .body
+        .push(ModuleItem::ModuleDecl(ModuleDecl::Import(import_decl)));
+}
+
+// fn generate_create_actor_options_interface() -> Box<TsInterfaceDecl> {
```
Comment (ilbertt) [inline rust/candid_parser/src/bindings/typescript_native/preamble/imports.rs:252]: Can we remove this comment?

File: rust/candid_parser/src/bindings/typescript_native/preamble/imports.rs
Comment (ilbertt) [inline rust/candid_parser/src/bindings/typescript_native/preamble/imports.rs:1]: Now that [`@icp-sdk/core`](https://www.npmjs.com/package/@icp-sdk/core) is out, we can change the imports to use the new package

**Review by ilbertt — COMMENTED** [review]
In the new generated bindings, we assume that the old bindings are still generated. I don't see a strong reason for that though.
We can generate the `_SERVICE` type in the new `ts-native-interface` wrapper. And we can add the old `createActor`'s function logic to the new `createActor` function in the `ts-native-wrapper`. What do you think?

In addition, we also assume that the old bindings are generated in a `declarations/*` folder, which may not be the case for many projects.

File: rust/candid_parser/src/bindings/typescript_native/new_typescript_native_types.rs [hunk 12]
```diff
@@ -0,0 +1,1107 @@
...
+
+                        let (span, syntax_field_ty) =
+                            find_field(top_level_nodes, syntax_fields, &f.id);
+
+                        // Create the __kind__ property
+                        let kind_prop = TsTypeElement::TsPropertySignature(TsPropertySignature {
+                            span,
+                            readonly: false,
+                            key: Box::new(Expr::Ident(Ident::new(
+                                "__kind__".into(),
```
Comment (ilbertt) [inline rust/candid_parser/src/bindings/typescript_native/new_typescript_native_types.rs:575]: We should declare a constant for this key and use it all over the `typescript_native` module:

```rust
const CANDID_VARIANT_KIND_KEY: &str = "__candid_variant_kind__"
```

In addition, I would use a harder-to-guess key like the one I suggested above to mitigate the risk of clashing with user-defined keys.

File: rust/candid_parser/src/bindings/typescript_native/preamble/options.rs [hunk 13]
```diff
@@ -0,0 +1,1001 @@
...
+                        op: BinaryOp::EqEqEq,
+                        left: Box::new(Expr::Member(MemberExpr {
+                            span: DUMMY_SP,
+                            obj: Box::new(Expr::Ident(Ident::new(
+                                "option".into(),
+                                DUMMY_SP,
+                                SyntaxContext::empty(),
+                            ))),
+                            prop: MemberProp::Ident(
+                                Ident::new("__kind__".into(), DUMMY_SP, SyntaxContext::empty())
```
Comment (ilbertt) [inline rust/candid_parser/src/bindings/typescript_native/preamble/options.rs:395]: Similar to https://github.com/dfinity/candid/pull/672#discussion_r2296717803, I would define a constant and use a harder-to-guess key:

```rust
const CANDID_OPTION_KIND_KEY: &str = "__candid_option_kind__"
```

File: rust/candid_parser/tests/assets/ok/typescript-interface/example.ts [hunk 14]
```diff
@@ -0,0 +1,930 @@
...
+}
+function record_opt_to_undefined<T>(arg: T | null): T | undefined {
+    return arg == null ? undefined : arg;
+}
+export type my_variant = {
+    /**
+     * Doc comment for my_variant field a
+    */
+    __kind__: "a";
+    a: {
```
Comment (ilbertt) [inline rust/candid_parser/tests/assets/ok/typescript-interface/example.ts:52]: I would set the comment on the actual field instead of the "internal" kind field:

```suggestion
    __kind__: "a";
    /**
     * Doc comment for my_variant field a
    */
    a: {
```

Same for the generated interface

File: rust/candid_parser/src/bindings/typescript_native/compile_interface.rs [hunk 15]
```diff
@@ -0,0 +1,184 @@
...
+                )
+            } else {
+                interface_actor_implementation(env, module, t, None, service_name, converter, span)
+            }
+        }
+        _ => {}
+    }
+}
+
+// Add actor implementation from service definition
```
Comment (ilbertt) [inline rust/candid_parser/src/bindings/typescript_native/compile_interface.rs:132]: There are some comments that we may want to fix/remove like this one. Can you give a check around the PR's code?

File: deny.toml
Comment (giorgiopiatti-caffeinated) [reply deny.toml:1]: Unicode-DFS-2016 seems to be an old name for Unicode-3.0, see [this](https://opensource.org/license/unicode-inc-license-agreement-data-files-and-software )

File: rust/candid/Cargo.toml
Comment (giorgiopiatti-caffeinated) [reply rust/candid/Cargo.toml:1]: Yes, let me clean this.

File: rust/candid_parser/src/bindings/typescript_native/comments.rs (hunk 10 above)
Comment (giorgiopiatti-caffeinated) [reply rust/candid_parser/src/bindings/typescript_native/comments.rs:48]: Yes, let me clean this.

File: rust/candid_parser/tests/assets/ok/typescript-interface/example.ts (hunk 14 above)
Comment (giorgiopiatti-caffeinated) [reply rust/candid_parser/tests/assets/ok/typescript-interface/example.ts:52]: Make sense also to me.
